File: core/src/main/java/net/dontdrinkandroot/wicket/component/basic/RepeatingList.java
/*
 * Copyright (C) 2012-2017 Philip Washington Sorst <philip@sorst.net>
 * and individual contributors as indicated
 * by the @authors tag.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dontdrinkandroot.wicket.component.basic;

import org.apache.wicket.markup.ComponentTag;
import org.apache.wicket.markup.html.panel.GenericPanel;
import org.apache.wicket.markup.repeater.RepeatingView;
import org.apache.wicket.model.IModel;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public abstract class RepeatingList<T> extends GenericPanel<T>
{
    public RepeatingList(String id)
    {
        super(id);
    }

    public RepeatingList(String id, IModel<T> model)
    {
        super(id, model);
    }

    @Override
    protected void onInitialize()
    {
        super.onInitialize();

        RepeatingView itemView = new RepeatingView("item");
        this.populateItems(itemView);
        this.add(itemView);
    }

    @Override
    protected void onComponentTag(ComponentTag tag)
    {
        super.onComponentTag(tag);
        if (!"ul".equals(tag.getName()) && !"ol".equals(tag.getName())) {
            this.findMarkupStream().throwMarkupException("Must be applied to ul or ol tag");
        }
    }

    protected abstract void populateItems(RepeatingView itemView);
}

File: core/src/main/kotlin/net/dontdrinkandroot/wicket/component/basic/RepeatingList.kt
package net.dontdrinkandroot.wicket.component.basic

import org.apache.wicket.markup.ComponentTag
import org.apache.wicket.markup.html.panel.GenericPanel
import org.apache.wicket.markup.repeater.RepeatingView
import org.apache.wicket.model.IModel

abstract class RepeatingList<T>(id: String?, model: IModel<T>? = null) : GenericPanel<T>(id, model) {

    override fun onInitialize() {
        super.onInitialize()
        val itemView = RepeatingView("item")
        populateItems(itemView)
        this.add(itemView)
    }

    override fun onComponentTag(tag: ComponentTag) {
        super.onComponentTag(tag)
        if ("ul" != tag.name && "ol" != tag.name) {
            findMarkupStream().throwMarkupException("Must be applied to ul or ol tag")
        }
    }

    protected abstract fun populateItems(itemView: RepeatingView)
}